src/main/java/spireMapOverhaul/BetterMapGenerator.java

```diff
@@ -36,12 +39,22 @@ private BetterMapGenerator() {
     }
 
     public ArrayList<ArrayList<MapRoomNode>> generate(Random rng, int width, int height, int pathDensity) {
+        if (Settings.isEndless && ActUtil.getRealActNum() == 1) {
+            //In endless mode, we let you encounter zones again each endless cycle
+            SpireAnniversary6Mod.currentRunSeenZones.clear();
+        }
+        else {
+            SpireAnniversary6Mod.currentRunSeenZones.addAll(activeZones.stream().map(z -> z.id).collect(Collectors.toList()));
+        }
+        mapGenLogger.info("Already seen zones: " + String.join(",", SpireAnniversary6Mod.currentRunSeenZones));
         MapPlanner planner;
         do {
             List<AbstractZone> possibleZones = new ArrayList<>();
-            for (AbstractZone zone : SpireAnniversary6Mod.allZones)
-                if (zone.canSpawn())
+            for (AbstractZone zone : SpireAnniversary6Mod.unfilteredAllZones) {
+                if (SpireAnniversary6Mod.currentRunAllZones.contains(zone.id) && !SpireAnniversary6Mod.currentRunSeenZones.contains(zone.id) && zone.canSpawn()) {
```

Comment:
Might be good to give people a utility method that gives you all zones that are not filtered for the current run, so that people don't have to do this logic themselves